db: range_tombstone_list: Avoid quadratic behavior when applying

tgrabiec opened the pull request:
Range tombstones are kept in memory (cache/memtable) in
range_tombstone_list. It keeps them deoverlapped, so applying a range
tombstone which covers many range tombstones will erase existing range
tombstones from the list. This operation needs to be exception-safe,
so range_tombstone_list maintains an undo log. This undo log will
receive a record for each range tombstone which is removed. For
exception safety reasons, before pushing an undo log entry, we reserve
space in the log by calling std::vector::reserve(size() + 1). This is
O(N) where N is the number of undo log entries. Therefore, the whole
application is O(N^2).

This can cause reactor stalls and availability issues when replicas
apply such deletions.

This patch avoids the problem by reserving exponentially increasing
amount of space. Also, to avoid large allocations, switches the
container to chunked_vector.

Fixes #11211

bhalevy commented on range_tombstone_list.cc:
```diff
@@ -373,15 +373,24 @@ range_tombstone_list::reverter::insert(range_tombstones_type::iterator it, range
+// Amortizes space expansion so that a series of N calls to amortized_reserve()
+// starting from an empty container takes O(N) time.
+template <typename T>
+void amortized_reserve(T& v) {
```
I think that it would be more elegant, and possibly more useful in the future
if `amortized_reserve` will be passed the desired capacity so it can be used with sizes
other than v.size()+1.

Otherwise, LGTM

tgrabiec replied:
Ok.

raphaelsc commented on range_tombstone_list.cc:
```diff
@@ -373,15 +373,24 @@ range_tombstone_list::reverter::insert(range_tombstones_type::iterator it, range
+template <typename T>
+void amortized_reserve(T& v) {
+    v.reserve(v.size() + std::max<size_t>(1, v.size()));
+}
```
why not log(N)? using growth factor of 1.5-2. although with chunked_vector, new room will only be made once last chunk is exhausted. looks like with this approach growth factor decreases as size increases, with capacity barely increasing the closer the factor is to 1. why not make the growth factor a constant?

tgrabiec replied:
The growth factor is constant, it's 2x. What do you mean by "why not log(N)"?

raphaelsc replied:
if size == 1, then capacity becomes 2, ok growth factor of 2.
if size == 2, then capacity becomes 4. ok.
but once size == 3, then capacity becomes 6. growth factor reduced to 1.5
....
if size == 20, then capacity is 40.
if size == 21, then capacity is 42. growth factor at this point is 1.05

if growth factor of *capacity* is 2, then we need log2(N) reallocations.

the problem seems size is being used as reference for calculating new capacity. is that what you really wanted to do?

raphaelsc replied:
https://godbolt.org/z/9ajc6YsWW

tgrabiec replied:
Right, will fix.

avikivity commented on range_tombstone_list.cc:
```diff
@@ -373,15 +373,24 @@ range_tombstone_list::reverter::insert(range_tombstones_type::iterator it, range
+// does not throw if element construction does not throw.
+// Amortizes space expansion so that a series of N calls to amortized_reserve()
+// starting from an empty container takes O(N) time.
+template <typename T>
```
Please add a constraint (or open code it). Plain templates that give no hint about the input should be avoided.

tgrabiec replied:
I'll just use chunked_vector<> then, it's not generically used yet.

avikivity commented on range_tombstone_list.cc:
```diff
@@ -373,15 +373,24 @@ range_tombstone_list::reverter::insert(range_tombstones_type::iterator it, range
+// starting from an empty container takes O(N) time.
+template <typename T>
+void amortized_reserve(T& v) {
+    v.reserve(v.size() + std::max<size_t>(1, v.size()));
```
Agree it should receive the desired capacity, and it should also examine the existing capacity.

I think as written it is still quadratic - an {amotized_reserve, push_back}+ sequence will call reserve() with arguments of 1, 2, 4, 6, 8, 10, 12, ... 2n since size() grows as n.

bhalevy replied:
Agreed.  `amortized_reserve` should examine `capacity()` and reserve only when required for the desired new capacity.

tgrabiec replied:
Right. Testing already showed a huge improvement due the switch to chunked_vector, for which reserve() is not O(N), but O(last_chunk_size).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1680/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1681/

tgrabiec commented:
In V2:

  - Moved amortized_reserve() to a header
  - Generalized amortized_reserve() with a concept
  - Fixed amortized_reserve() to really amortize reserve() calls
  - Added a unit test for amortized_reserve()
  - Added desired_size parameter to amortized_reserve() to mimic reserve()

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1683/

raphaelsc commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1683/

I think you missed git add on new header for amortized reserve

tgrabiec commented:
In V3:

   - Added missing header with amortized_reserve()

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1684/

denesb commented on range_tombstone_list.cc:
```diff
@@ -375,13 +376,13 @@ range_tombstone_list::reverter::insert(range_tombstones_type::iterator it, range
 range_tombstone_list::range_tombstones_type::iterator
 range_tombstone_list::reverter::erase(range_tombstones_type::iterator it) {
-    _ops.reserve(_ops.size() + 1);
+    amortized_reserve(_ops, _ops.size() + 1);
```
Isn't this the same strategy `std::vector` follows by default? I think this code would not have been quadratic without the manual reserve call.

denesb replied:
Would just removing the reserve altogether also fix this?

tgrabiec replied:
The reason we need to reserve() is so that push_back() does not throw.

denesb replied:
See @kbr-'s comment below.

kbr- commented:
@tgrabiec if I understand correctly, the entire point of this `reserve` is to make sure space is allocated before you construct `erase_undo_op`, because the destructor of `erase_undo_op` has a side effect, and you don't want this side effect to happen if vector allocation fails.

You can achieve your goal without `reserve` by not constructing the element before you call `emplace_back`. Right now it's first constructed using the `range_tombstone_entry&` constructor (the `erase_undo_op(*it)` expression), then `emplace_back` is called which move-constructs the element again inside the vector. BTW. if you do it like that, there's no reason to use `emplace_back`, `push_back` would give the same effect; the whole point of `emplace_back` is avoiding the move constructor.

Instead, you could construct the element in-place:
```cpp
_ops.emplace_back(std::in_place_type<erase_undo_op>, *it);
```
(see https://en.cppreference.com/w/cpp/utility/variant/variant)

With this, the `erase_undo_op` will be constructed once after the allocation.

tgrabiec commented:
Good suggestion @kbr-, I'll use it. 

> BTW. if you do it like that, there's no reason to use emplace_back, push_back would give the same effect; the whole point of emplace_back is avoiding the move constructor.

Still, with push_back() we would call the move constructor two times instead of once. This is because first we construct the variant, then move the variant.

kbr- commented:
> Still, with push_back() we would call the move constructor two times instead of once. This is because first we construct the variant, then move the variant.

Right.
